plugins/experimental/ssl_session_reuse/src/publish.cc

```diff
@@ -0,0 +1,401 @@
+                re.m_hostname.c_str(), re.m_port, static_cast<int>(my_id), (i + 1));
+        continue;
+      }
+    }
+
+    current_reply = static_cast<redisReply *>(::redisCommand((*ctx), "PUBLISH %s %s", msg.channel.c_str(), msg.data.c_str()));
+    if (!current_reply) {
+      TSError("RedisPublisher::send_publish.Unable to get a reply from the server for publish.ThreadId: %d Try: %d",
+              static_cast<int>(my_id), (i + 1));
+      teardown_connection(*ctx);
+      *ctx = NULL;
+
+    } else if (REDIS_REPLY_ERROR == current_reply->type) {
+      TSError("RedisPublisher::send_publish.Server responded with error for publish.ThreadId: %d Try: %d", static_cast<int>(my_id),
+              i + 1);
+      clear_reply(current_reply);
+      current_reply = NULL;
+      teardown_connection(*ctx);
+      *ctx = NULL;
+    } else {
+      break;
+    }
+  }
+
+  return current_reply;
+}
+
+void
+RedisPublisher::clear_reply(::redisReply *reply)
+{
+  if (reply) {
+    ::freeReplyObject(reply);
+  }
+}
+
+void
+RedisPublisher::runWorker()
+{
+  ::pthread_mutex_lock(&m_endpointIndexMutex);
+  RedisEndpoint &my_endpoint(m_redisEndpoints[m_endpointIndex]);
+
+  if (static_cast<int>(m_redisEndpoints.size() - 1) == m_endpointIndex) {
+    m_endpointIndex = 0;
+  } else {
+    ++m_endpointIndex;
+  }
+  ::pthread_mutex_unlock(&m_endpointIndexMutex);
+
+  ::pthread_t my_id = ::pthread_self();
+  ::redisContext *my_context(setup_connection(my_endpoint));
+  ::redisReply *current_reply(NULL);
+
+  while (true) {
+    ::sem_post(&m_workerSem);
+    int readyWorkers = 0;
+    ::sem_getvalue(&m_workerSem, &readyWorkers);
+
+    // LOGDEBUG(PUB, "RedisPublisher::runWorker.Ready worker count: " << readyWorkers);
+
+    ::pthread_mutex_lock(&m_messageQueueMutex);
+
+    if (m_messageQueue.empty()) {
+      ::sem_wait(&m_workerSem);
+      ::pthread_mutex_unlock(&m_messageQueueMutex);
+      usleep(1000);
+      continue;
+    }
+
+    Message current_message(m_messageQueue.front());
+
+    if (!current_message.cleanup) {
+      m_messageQueue.pop_front();
+    }
+
+    ::pthread_mutex_unlock(&m_messageQueueMutex);
+    ::sem_wait(&m_workerSem);
+
+    if (current_message.cleanup) {
+      TSDebug(PLUGIN, "RedisPublisher::runWorker.Thread id: %d received the cleanup message. exiting!", static_cast<int>(my_id));
+      break;
+    }
+    current_reply = send_publish(&my_context, my_endpoint, current_message);
+
+    if (!current_reply) {
+      current_message.hosts_tried.insert(my_endpoint);
+      if (current_message.hosts_tried.size() < m_redisEndpoints.size()) {
+        // all endpoints are not tried
+        // someone else might be able to transmit
+        ::pthread_mutex_lock(&m_messageQueueMutex);
+
+        if (!m_messageQueue.front().cleanup) {
+          m_messageQueue.push_front(current_message);
+        }
+
+        ::pthread_mutex_unlock(&m_messageQueueMutex);
+      }
+    }
```

Comment:
Move:

if (!current_message.cleanup) {
      m_messageQueue.pop_front();
}

here if current_message is changed to a reference.